commands/controllerremote.go

```diff
@@ -0,0 +1,328 @@
+			if err := os.Remove(addr); err != nil && !os.IsNotExist(err) { // avoid EADDRINUSE
+				return err
+			}
+			defer func() {
+				if err := os.Remove(addr); err != nil {
+					logrus.Errorf("failed to clean up socket %q: %v", addr, err)
+				}
+			}()
+			logrus.Infof("starting server at %q", addr)
+			l, err := net.Listen("unix", addr)
+			if err != nil {
+				return err
+			}
+			rpc := grpc.NewServer()
+			controllerapi.RegisterControllerServer(rpc, b)
+			doneCh := make(chan struct{})
+			errCh := make(chan error, 1)
+			go func() {
+				defer close(doneCh)
+				if err := rpc.Serve(l); err != nil {
+					errCh <- fmt.Errorf("error on serving via socket %q: %w", addr, err)
+				}
+				return
+			}()
+			var s os.Signal
+			sigCh := make(chan os.Signal, 1)
+			signal.Notify(sigCh, os.Interrupt)
+			select {
+			case s = <-sigCh:
+				logrus.Debugf("got signal %v", s)
+			case err := <-errCh:
+				return err
+			case <-doneCh:
+			}
+			return nil
+
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&serverConfigPath, "config", "", "Specify buildx server config file")
+	return cmd
+}
+
+func getLogFilePath(configPath string) (string, error) {
+	config, err := getConfig(configPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to get config")
+	}
+	logFile := config.LogFile
+	if logFile == "" {
+		root, err := prepareRootDir(config)
+		if err != nil {
+			return "", err
+		}
+		logFile = filepath.Join(root, "log")
+	}
+	return logFile, nil
+}
+
+func getConfig(configPath string) (*serverConfig, error) {
+	var defaultConfigPath bool
+	if configPath == "" {
+		defaultRoot, err := rootDataDir()
+		if err != nil {
+			return nil, err
+		}
+		configPath = filepath.Join(defaultRoot, "config.toml")
+		defaultConfigPath = true
+	}
+	var config serverConfig
+	tree, err := toml.LoadFile(configPath)
+	if err != nil && !(os.IsNotExist(err) && defaultConfigPath) {
+		return nil, fmt.Errorf("failed to load config file %q", configPath)
+	} else if err == nil {
+		if err := tree.Unmarshal(&config); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal config file %q", configPath)
+		}
+	}
+	return &config, nil
+}
+
+func prepareRootDir(config *serverConfig) (string, error) {
+	rootDir := config.Root
+	if rootDir == "" {
+		var err error
+		rootDir, err = rootDataDir()
+		if err != nil {
+			return "", err
+		}
+	}
+	if rootDir == "" {
+		return "", fmt.Errorf("buildx root dir must be determined")
+	}
+	if err := os.MkdirAll(rootDir, 0700); err != nil {
+		return "", err
+	}
+	serverRoot := filepath.Join(rootDir, "shared")
+	if err := os.MkdirAll(serverRoot, 0700); err != nil {
+		return "", err
+	}
+	return serverRoot, nil
+}
+
+func rootDataDir() (string, error) {
+	if xdh := os.Getenv("XDG_DATA_HOME"); xdh != "" {
+		return filepath.Join(xdh, "buildx"), nil
+	}
+	home := os.Getenv("HOME")
+	if home == "" {
+		return "", fmt.Errorf("environment variable HOME is not set")
+	}
+	return filepath.Join(home, ".local/share/buildx"), nil
+}
```

Comment:
I wonder if we should not just use the current buildx config dir: https://github.com/docker/buildx/blob/b242e3280b0c682bcf9aa7e10cffcfc1099f80f2/util/confutil/config.go#L17-L26

And create a `controller` folder in it?

Reply:
Thanks for the review. Fixed.